write-fonts/src/pens.rs

```diff
@@ -0,0 +1,139 @@
+//! Pen implementations based on <https://github.com/fonttools/fonttools/tree/main/Lib/fontTools/pens>
+
+use std::mem;
+
+use font_types::Pen;
+use kurbo::{Affine, BezPath, Point};
+
+/// A pen that transforms params using [kurbo::Affine].
+pub struct TransformPen<'a, T: Pen> {
+    inner_pen: &'a mut T,
+    transform: Affine,
+}
+
+impl<'a, T: Pen> TransformPen<'a, T> {
+    pub fn new(inner_pen: &'a mut T, transform: Affine) -> TransformPen<'a, T> {
+        TransformPen {
+            inner_pen,
+            transform,
+        }
+    }
+
+    fn map_point(&self, x: f32, y: f32) -> (f32, f32) {
+        let pt = self.transform
+            * Point {
+                x: x as f64,
+                y: y as f64,
+            };
+        (pt.x as f32, pt.y as f32)
+    }
+}
+
+impl<'a, T: Pen> Pen for TransformPen<'a, T> {
+    fn move_to(&mut self, mut x: f32, mut y: f32) {
+        (x, y) = self.map_point(x, y);
+        self.inner_pen.move_to(x, y);
+    }
+
+    fn line_to(&mut self, mut x: f32, mut y: f32) {
+        (x, y) = self.map_point(x, y);
+        self.inner_pen.line_to(x, y);
+    }
+
+    fn quad_to(&mut self, mut cx0: f32, mut cy0: f32, mut x: f32, mut y: f32) {
+        (cx0, cy0) = self.map_point(cx0, cy0);
+        (x, y) = self.map_point(x, y);
+        self.inner_pen.quad_to(cx0, cy0, x, y);
+    }
+
+    fn curve_to(
+        &mut self,
+        mut cx0: f32,
+        mut cy0: f32,
+        mut cx1: f32,
+        mut cy1: f32,
+        mut x: f32,
+        mut y: f32,
+    ) {
+        (cx0, cy0) = self.map_point(cx0, cy0);
+        (cx1, cy1) = self.map_point(cx1, cy1);
+        (x, y) = self.map_point(x, y);
+        self.inner_pen.curve_to(cx0, cy0, cx1, cy1, x, y);
+    }
+
+    fn close(&mut self) {
+        self.inner_pen.close();
+    }
+}
+
+pub struct BezPathPen {
+    path: BezPath,
+}
+
+fn pt(x: f32, y: f32) -> Point {
+    Point {
+        x: x as f64,
+        y: y as f64,
+    }
+}
+
+impl BezPathPen {
+    pub fn new() -> BezPathPen {
+        BezPathPen {
+            path: BezPath::new(),
+        }
+    }
+
+    pub fn take_path(&mut self) -> BezPath {
```

Comment:
Unless you really need to hold on to the pen, the common name for this is `into_inner` which takes self by value and just returns the path. It might also make sense to add a `From` impl.

Reply:
Good catch, so updated.